crates/biome_js_analyze/src/lint/nursery/use_starts_ends_with.rs

```diff
@@ -0,0 +1,409 @@
+use biome_analyze::{Ast, Rule, RuleDiagnostic, context::RuleContext, declare_lint_rule};
+use biome_console::markup;
+use biome_diagnostics::Severity;
+use biome_js_syntax::{
+    AnyJsExpression, JsBinaryExpression, JsBinaryOperator, JsCallExpression,
+    JsComputedMemberExpression, JsRegexLiteralExpression, JsStaticMemberExpression,
+};
+use biome_rowan::{AstNode, AstSeparatedList, declare_node_union};
+
+declare_lint_rule! {
+    /// Enforce using `String.startsWith()` and `String.endsWith()` over more complex alternatives.
+    ///
+    /// There are multiple ways to verify if a string starts or ends with a certain string,
+    /// such as `foo.indexOf('bar') === 0` or `foo.slice(0, 3) === 'bar'`.
+    ///
+    /// This rule is aimed to enforce a consistent and more readable style using `startsWith()` and `endsWith()` methods.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "hello";
+    /// foo[0] === 'h';
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "hello";
+    /// foo.charAt(0) === 'h';
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "hello";
+    /// foo.indexOf('bar') === 0;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = "hello";
+    /// /^bar/.test(foo);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const foo = "hello";
+    /// foo.startsWith('h');
+    /// ```
+    ///
+    /// ```js
+    /// const foo = "hello";
+    /// foo.endsWith('o');
+    /// ```
+    ///
+    pub UseStartsEndsWith {
+        version: "next",
+        name: "useStartsEndsWith",
+        language: "js",
+        recommended: false,
+        severity: Severity::Warning,
+    }
+}
+
+declare_node_union! {
+    pub AnyStringCheckExpression = JsBinaryExpression | JsCallExpression
+}
+
+#[derive(Debug, Clone)]
+pub enum StartsEndsWithVariant {
+    StartsWith,
+    EndsWith,
+}
+
+#[derive(Debug, Clone)]
+pub struct UseStartsEndsWithState {
+    pub variant: StartsEndsWithVariant,
+    pub pattern: String,
+}
+
+impl Rule for UseStartsEndsWith {
+    type Query = Ast<AnyStringCheckExpression>;
```

Comment:
The source rule uses type information to only make these suggestions for checks against actual `string`s. For example `foo[0] === "b"` could have `foo` be an array, and this rule would flag it as a false positive.